compiler/next/lib/resolution/resolution-types.cpp

```diff
@@ -17,11 +17,235 @@
+        mappingIsValid = false;
+        failingActualIdx = actualIdx;
+        // TODO: track failure type for error messages
+        return false;
+      }
+    }
+    actualIdx++;
+  }
+
+  // Fill in unmatched formals in sequence with the remaining actuals.
+  // Record successful substitutions
+  formalIdx = 0;
+  actualIdx = 0;
+  for (const CallInfoActual& actual : call.actuals) {
+    if (actual.byName.isEmpty()) {
+      // Skip any formals already matched to named arguments
+      while (byFormalIdx[formalIdx].actualIdx >= 0) {
+        if (formalIdx + 1 >= byFormalIdx.size()) {
+          // too many actuals
+          mappingIsValid = false;
+          failingActualIdx = actualIdx;
+          return false;
+        }
+        formalIdx++;
+      }
+      assert(formalIdx < byFormalIdx.size());
+
+      // TODO: special handling for operators
+
+      FormalActual& entry = byFormalIdx[formalIdx];
+      entry.hasActual = true;
+      entry.actualIdx = actualIdx;
+      entry.actualType = actual.type;
+      actualIdxToFormalIdx[actualIdx] = formalIdx;
+    }
+    actualIdx++;
+  }
+
+  // Make sure that any remaining formals are matched by name
+  // or have a default value.
+  while (formalIdx < byFormalIdx.size()) {
+    if (byFormalIdx[formalIdx].actualIdx < 0 &&
+        untyped->formals[formalIdx]->initExpression() == nullptr) {
+      // formal was not provided and there is no default value
+      mappingIsValid = false;
+      failingFormalIdx = formalIdx;
+      return false;
+    }
+    formalIdx++;
+  }
+
+  mappingIsValid = true;
+  return true;
+}
+
+FormalActualMap FormalActualMap::build(const UntypedFnSignature* sig,
+                                       const CallInfo& call) {
+
+  FormalActualMap ret;
+  bool ok = ret.computeAlignment(sig, nullptr, call);
+  if (!ok) {
+    ret.mappingIsValid = false;
+  }
+  return ret;
+}
+
+FormalActualMap FormalActualMap::build(const TypedFnSignature* sig,
+                                       const CallInfo& call) {
+  FormalActualMap ret;
+  bool ok = ret.computeAlignment(sig->untypedSignature, sig, call);
+  if (!ok) {
+    ret.mappingIsValid = false;
+  }
+  return ret;
+}
+
+std::string TypedFnSignature::toString() const {
+  std::string ret = untypedSignature->functionId.toString();
+  ret += "(";
+  for (size_t i = 0; i < untypedSignature->formals.size(); i++) {
+    if (i != 0) {
+      ret += ", ";
+    }
+    ret += untypedSignature->formals[i]->name().c_str();
+    ret += " : ";
+    ret += formalTypes[i].toString();
+  }
+  ret += ")";
+
+  return ret;
+}
+
+
+void PoiInfo::accumulate(const PoiInfo& addPoiInfo) {
+  poiFnIdsUsed.insert(addPoiInfo.poiFnIdsUsed.begin(),
+                      addPoiInfo.poiFnIdsUsed.end());
+}
+
+// this is a resolved function
+// check is a not-yet-resolved function
+bool PoiInfo::canReuse(const PoiInfo& check) const {
+  assert(resolved && !check.resolved);
+
+  return false; // TODO -- consider function names etc -- see PR #16261
```

Comment:
This is basically a stub, right?  I'm not sure I fully understand its utility in its current form otherwise

Reply:
Yes, that's right. I wanted to demonstrate that we could get a reasonable structure of the resolver. It doesn't yet implement the checks from #16261. That is future work.